docs/codeql/codeql-language-guides/customizing-library-models-for-java.rst

```diff
@@ -0,0 +1,403 @@
+- **sinkModel**\(package, type, subtypes, name, signature, ext, input, kind, provenance)
+- **summaryModel**\(package, type, subtypes, name, signature, ext, input, output, kind, provenance)
+- **neutralModel**\(package, type, name, signature, provenance)
+
+TODO: Link or inline documentation on how to add dataextensions.
+Are we going for extensions packs as the recommended default?
+If yes, then we probably need to elaborate with a concrete example.
+
+In the sections below, we will show by example how to add tuples to the different extension points.
+The extension points are used to customize and improve the existing dataflow queries, by providing sources, sinks, and flow through for library elements.
+The **Reference material** section will in more detail describe the *mini DSLs* that are used to comprise a model definition for each extension point.
+
+Example: Taint sink in the **java.sql** package.
+------------------------------------------------
+
+In this example we will see, how to define the argument of the **execute** method as a SQL injection sink.
+This is the **execute** method in the **Statement** class, which is located in the **java.sql** package.
+Please note that this sink is already added to the CodeQL Java analysis.
+
+.. code-block:: java
+
+   public static void taintsink(Connection conn, String query) throws SQLException {
+       Statement stmt = conn.createStatement();
+       stmt.execute(query); // The argument to this method is a SQL injection sink.
+   }
+
+This can be achieved by adding the following data extension.
+
+.. code-block:: yaml
+
+   extensions:
+     - addsTo:
+         pack: codeql/java-all
+         extensible: sinkModel
+       data:
+         - ["java.sql", "Statement", True, "execute", "(String)", "", "Argument[0]", "sql", "manual"]
+
+Reasoning:
+
+Since we are adding a new sink, we need to add a tuple to the **sinkModel** extension point.
+The first five values are used to identify the method (callable) which we are defining a sink on.
+
+- The first value **java.sql** is the package name.
+- The second value **Statement** is the class (type) name.
+- The third value **True** is flag indicating, whether the sink also applies to all overrides of the method.
+- The fourth value **execute** is the method name.
+- The fifth value **(String)** is the method input type signature.
+
+For most practical purposes the sixth value is not relevant.
+The remaining values are used to define the **access path**, the **kind**, and the **provenance** (origin) of the sink.
+
+- The seventh value **Argument[0]** is the **access path** to the first argument passed to the method, which means that this is the location of the sink.
+- The eighth value **sql** is the kind of the sink. The sink kind is used to define the queries where the sink is in scope. In this case  - the SQL injection queries.
+- The ninth value **manual** is the provenance of the sink, which is used to identify the origin of the sink.
```

Comment:
Presumably, all hand written sinks should be `manual`. What other values can there be?